source/agora/consensus/EnrollmentManager.d

```diff
@@ -699,6 +734,100 @@ public class EnrollmentManager
+
+        Returns:
+            true if revealing a pre-image is needed
+
+    ***************************************************************************/
+
+    public bool needRevealPreimage (ulong height) @safe nothrow
+    {
+        if (height >= this.next_reveal_height)
+            return true;
+        else
+            return false;
+    }
+
+    /***************************************************************************
+
+        Increase the next reveal height by the revelation period
+
+    ***************************************************************************/
+
+    public void increaseNextRevealHeight () @safe nothrow
+    {
+        ulong next_height = this.getNextRevealHeight();
+        if (this.next_reveal_height < ulong.max)
+            this.setNextRevealHeight(next_height + PreimageRevealPeriod);
+    }
+
+    /***************************************************************************
+
+        Get the next block height to reveal a pre-image
+
+        Returns:
+            the next block height to reveal a pre-image. if any problem in
+            getting the value, it returns the MAX ulong value.
+
+    ***************************************************************************/
+
+    private ulong getNextRevealHeight () @safe nothrow
+    {
+        ulong next_height = ulong.max;
+        try
+        {
+            () @trusted {
+                auto results = this.db.execute("SELECT val FROM node_enroll_data " ~
+                    "WHERE key = ?", "next_reveal_height");
+                if (!results.empty)
+                    next_height = results.oneValue!(size_t);
+            }();
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error {}", ex);
+        }
+        return next_height;
+    }
+
+    /***************************************************************************
+
+        Set the next block height to reveal a pre-image
+
+        Params:
+            height = the next block height to reveal a pre-image
+
+    ***************************************************************************/
+
+    private void setNextRevealHeight (ulong height) @safe nothrow
+    {
+        try
+        {
+            () @trusted {
+                auto results = this.db.execute("SELECT EXISTS(SELECT 1 FROM " ~
+                    "node_enroll_data WHERE key = ?)", "next_reveal_height");
+                if (results.oneValue!(bool))
+                    this.db.execute("UPDATE node_enroll_data SET val = ? " ~
+                        "WHERE key = ?", height, "next_reveal_height");
+                else
+                    this.db.execute("INSERT INTO node_enroll_data (key, val)" ~
+                        " VALUES (?, ?)", "next_reveal_height", height);
+
+                this.next_reveal_height = height;
+            }();
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error {}", ex);
```

Comment:
Same problem here. We are logging errors and then proceeding as if there was no error.

Reply:
> Same problem here. We are logging errors and then proceeding as if there was no error.

There needs to add the code of `assert(0)` in a related PR.